set.mm

```diff
@@ -18668,6 +18668,17 @@ This axiom scheme is logically redundant (see ~ ax12w ) but is used as an
       ( wex ex exlimd mpd ) ABDICGABCDEFABCHJKL $.
   $}
 
+  ${
+    exlimimdd.1 $e |- F/ x ph $.
+    exlimimdd.2 $e |- F/ x ch $.
+    exlimimdd.3 $e |- ( ph -> E. x ps ) $.
+    exlimimdd.4 $e |- ( ph -> ( ps -> ch ) ) $.
+    $( Existential elimination rule of natural deduction.  (Contributed by ML,
+       17-Jul-2020.) $)
+    exlimimdd $p |- ( ph -> ch ) $=
+      ( imp exlimdd ) ABCDEFGABCHIJ $.
+  $}
```

Comment:
This theorem was moved to main because of the rule "theorem in a mathbox is used in a proof within another one's mathbox"). Furthermore, 7 proofs can be shortened by this theorem. Therefore, I would like to keep it here.

Reply:
Ok.
Is this rule written somewhere ? It contradicts the rule that a theorem should "pay for itself". In the present case: if exliimdd shortens 8 proofs, each by say 10 bytes (since it simply replaces "imp exlimdd" by "exlimimdd"), then it saves 80 bytes, but its statement+proof is more than 80 bytes).
And again: it favours a proof strategy different from the favored one.

Reply:
To be clearer on the last point: compare the following two proofs of the theorem `|- a curried hypothesis => |- a curried conclusion`:
```
a curried hypothesis
uncurry it
...
manipulations on uncurried expressions
...
curry it
a curried conclusion
```
versus
```
a curried hypothesis
...
manipulations on curried expressions
...
a curried conclusion
```
Of course, the second proof saves two (or maybe a bit more) lines, but we now need to have a library of lemmas for manipulating curried expressions (in addition to the already present library of lemmas for manipulating uncurried expressions), and these libraries are large.
This is the same as
```
a hypothesis with =/=
convert =/= to -. =
...
manipulations on expressions with =
...
convert -. = to =/=
a conclusion with =/=
```
versus
```
a hypothesis with =/=
...
manipulations on expressions with =/=
...
a conclusion with =/=
```
Again, to save a few proof lines, we need a whole new library of lemmas for inequalities, paralleling and almost duplicating the existing lemmas for equalities.

Reply:
It was a proposal by @david-a-wheeler in https://groups.google.com/g/metamath/c/JsNryVZhFRU/m/8RfDMaHtCwAJ in 2019:

```
As a general rule, if someone else finds something useful in a mathbox,
then we promote it to the main body (unless there's some serious
known problem). Stuff in the main body is much easier to reuse
and tends to be organized more logically. 
```
which was accepted by Norm, and since then practiced regularily (since my mathbox is almost at the end of set.mm, I often find theorems in previous mathboxes which I can use). I moved them to main, and Norm accepted it almost always without looking at the potential to shorten other proofs.

Reply:
> Ok. Is this rule written somewhere ? 

... and it is written in set.mm itself (see section "mathbox guidelines"):

```
5.  Mathboxes must be independent from one another (checked by "verify
       markup *").  If you need a theorem from another mathbox, typically it is
       moved to the main part of set.mm.  New users should consult with more
       experienced users before doing this.  
```

Reply:
> Plus, compared to exlimd, the only difference is that one of its hypotheses is in curried form, but when using deduction style we favor uncurried form. So making it available could even favor a less orthodox proof style.

For this I also have to ask: Is this rule written somewhere? That means, where is it written that "using deduction style we favor uncurried form"? Just to be clear: "uncurried" (usually used for functions) in the context of propositional calculus means `( ( ph /\ ps ) -> ch )` ("imported" form in the meaning of ~imp), whereas "curried" means `( ph -> ( ps -> ch ) )` ("exported" form in the meaning of ~ex), doesn't it?

Following the rule (1) for a theorem in deduction form:

`every hypothesis ($e statement) is either (1) an implication with the same antecedent as the conclusion, or ...`

More formal: If the conclusion is in the form ` ( ph -> ps ) ` (has antecedent ` ph `), then a hypothesis must have the same antecedent ` ph `, meaning must be in the form   ` ( ph -> ta ) `.

By this rule, the "curried" form `( ph -> ( ps -> ch ) )` would be in deduction form, and not the uncurried one.

I think we have to continue the discussion somewhere else (in a new GitHub issue, or in the Google group): not only about if the curried or uncurried version is preferred for the deduction style, but also to reduce the number of theorems eliminating lemmata for special definitions like =/= whic can be transformed in other theorems easily (as you described  in your comment yesterday).

Reply:
I didn't say it is a rule, but I explained above why it seems better to have such a convention, to avoid duplicating "library theorems". The idea of favoring the uncurried form (which is indeed as you wrote) in proofs was voiced on a few occasions on the discussion group (notably by @digama0). Indeed this should be discussed more widely on the discussion group. I'll try to open a topic.
(The "ok" above meant that this discussion does not block from merging this PR.)